crates/ra_hir/src/impl_block.rs

```diff
@@ -0,0 +1,169 @@
+use std::sync::Arc;
+use rustc_hash::FxHashMap;
+
+use ra_syntax::ast::{self, AstNode};
+use ra_db::{LocationIntener, Cancelable};
+
+use crate::{
+    Crate, DefId, DefLoc, DefKind, SourceItemId, SourceFileItems,
+    Module, Function,
+    db::HirDatabase,
+    type_ref::TypeRef,
+    arena::{Arena, Id},
+    module::{ModuleSourceNode},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ImplBlock {
+    crate_impl_blocks: Arc<CrateImplBlocks>,
+    impl_id: ImplId,
+}
+
+impl ImplBlock {
+    pub(crate) fn containing(
+        crate_impl_blocks: Arc<CrateImplBlocks>,
+        def_id: DefId,
+    ) -> Option<ImplBlock> {
+        let impl_id = *crate_impl_blocks.impls_by_def.get(&def_id)?;
+        Some(ImplBlock {
+            crate_impl_blocks,
+            impl_id,
+        })
+    }
+
+    fn impl_data(&self) -> &ImplData {
+        &self.crate_impl_blocks.impls[self.impl_id]
+    }
+
+    pub fn target(&self) -> &TypeRef {
+        &self.impl_data().impl_for
```

Comment:
We use two names (`target` and `impl_for`) for the same thing. We will use `ImplBlock` for trait impls as well, right? Then perpahs `fn target_type() -> &TypeRef` and `fn target_trait() ->Option<&TraitRef>`?

Reply:
yeah, I actually thought about renaming `impl_for`, but forgot again ;)